launch_ros/launch_ros/utilities/to_parameters_list.py

```diff
@@ -28,25 +29,69 @@
 from ..parameters_type import EvaluatedParameters
 
 
+def __format_dict(dictionary, keys, formatted_dict):
+    """
+    Combine namespaces and their node names into one key and remove the "ros__parameters" key.
+
+    Return an empty dict if the "ros__parameters" key is not found.
+    """
+    for key, value in dictionary.items():
+        if key == 'ros__parameters':
+            node_name = '/'.join(keys)
+            formatted_dict[node_name] = value
+            return formatted_dict
+        elif type(value) is dict:
+            keys.append(key.lstrip('/'))
+            formatted_dict = __format_dict(value, keys, formatted_dict)
+            keys = []
+    return formatted_dict
+
+
 def to_parameters_list(
     context: LaunchContext,
+    node_name: str,
+    namespace: str,
     evaluated_parameters: EvaluatedParameters
 ) -> List[rclpy.parameter.Parameter]:
     """
     Transform evaluated parameters into a list of rclpy.parameter.Parameter objects.
 
     :param context: to carry out substitutions during normalization of parameter files.
         See `normalize_parameters()` documentation for further reference.
+    :param node_name: node name
+    :param namespace: node namespace
     :param parameters: parameters as either paths to parameter files or name/value pairs.
         See `evaluate_parameters()` documentation for further reference.
     :returns: a list of parameters
     """
     parameters = []  # type: List[rclpy.parameter.Parameter]
+    node_name = node_name.lstrip('/')
+    if namespace:
+        namespace = namespace.lstrip('/')
+        node_name = f'{namespace}/{node_name}'
+
+    warned_once = False
     for params_set_or_path in evaluated_parameters:
         if isinstance(params_set_or_path, pathlib.Path):
             with open(str(params_set_or_path), 'r') as f:
+                param_dict = yaml.safe_load(f)
+                formatted_dict = __format_dict(param_dict, keys=[], formatted_dict={})
+
+                if formatted_dict:
+                    param_dict.clear()
+                    if '**' in formatted_dict:
+                        param_dict = formatted_dict['**']
+                    if node_name in formatted_dict:
+                        param_dict.update(formatted_dict[node_name])
+                    if not warned_once and not node_name:
+                        warnings.warn(
+                            'node name not provided to launch; parameter files will not apply',
+                            UserWarning
+                        )
+                        warned_once = True
+
                 params_set = evaluate_parameter_dict(
-                    context, normalize_parameter_dict(yaml.safe_load(f))
+                    context, normalize_parameter_dict(param_dict)
```

Comment:
@jacobperron we should skip this if `param_dict` is empty (when a parameter file is given a neither wildcard nor node name yield any parameters).

Reply:
Donee: 9de0767